### ISSUE TITLE
The gap between continued training and one-time training

### ISSUE THREAD

**Issue by CarlanLark** [microsoft/LightGBM#2204]
## Environment info

Operating System: Win 10

CPU/GPU model: CPU

C++/Python/R version: Python 3.6.3

LightGBM version or commit hash: 2.2.4

## Error message

I use init_score() to implement continued training, but the results of continued training and one-time train are different(a significant gap). I do 10 iterations both training and then calculate the gap. The absolute mean gap is approximately 25,000 while the magnitude of prediction is 10^5. So what causes the gap?
Here's the link of the data I use [https://www.kaggle.com/harlfoxem/housesalesprediction](url)
## Reproducible examples

```
import pandas as pd
import numpy as np
import lightgbm as lgb

#params for lgb
params = {
        'boosting_type': 'gbdt', 
        'objective': 'regression_l1', 
        'metric': {'l1'},
        'num_leaves': 31,
        'max_depth': 6,
        'min_data_in_leaf': 20,
        'learning_rate': 0.05,
        'feature_fraction': 0.8,
        'bagging_fraction': 0.95,
        #'bagging_freq': 5,
        'lambda_l1': 1,
        'lambda_l2': 0.001, 
        'min_gain_to_split': 0.2,
        'verbose': 5,
        'is_unbalance': True, 
        }

#preprocessing
df = pd.read_csv('kc_house_data.txt')
y = df[['price']]
df.drop(['id', 'price', 'date'], axis = 1, inplace = True)
x = df
lgb_train = lgb.Dataset(x, label = y)

#one-time training
gbm1 = lgb.train(params, lgb_train, num_boost_round = 10)
res1 = gbm1.predict(x)

#continued training
train_init_score = np.zeros(len(x))
for i in range(10):
    
    gbm2 = lgb.train(params, lgb_train, num_boost_round = 1)
    
    preds_train = gbm2.predict(x, raw_score = True)
    train_init_score += preds_train
    lgb_train.set_init_score(train_init_score)
    
res2 = train_init_score

#gap
gap = np.mean(np.abs(res1 - res2))
```
[kc_house_data.txt](https://github.com/microsoft/LightGBM/files/3258079/kc_house_data.txt)

**Comment by guolinke** [comment]
when with `init_score`, the `boost_from_average` will be disabled.
you can set `boost_from_average=false` for the fair comparison.
https://github.com/microsoft/LightGBM/blob/b1e5a843ab57c64b01c8896ef3cda6fdf907d417/src/boosting/gbdt.cpp#L310

**Comment by CarlanLark** [comment]
> when with init_score, the boost_from_average will be disabled.
> you can set boost_from_average=false for the fair comparison.
> 
>   
>     
>       LightGBM/src/boosting/gbdt.cpp
>     
>     
>          Line 310
>       in
>       b1e5a84
>     
>     
>     
>     
> 
>         
>           
>            if (models_.empty() && !train_score_updater_->has_init_score() && objective_function_ != nullptr) {

Thank you for your reply. But after setting `boost_from_average = false`, all predictions become zero.

**Comment by guolinke** [comment]
Sorry, set boost_from_average=true is okay in your case.

I saw the version you used is 2.2.3, we fix several bugs for L1 objective after that. could you try the latest master code? 

If this still cannot work, could you upload the data here? so that I can investigate it. Thanks.

**Comment by CarlanLark** [comment]
I updated my lightgbm to the latest version. But the gap still exists. I have uploaded the data [kc_house_data.txt](https://github.com/microsoft/LightGBM/files/3258079/kc_house_data.txt)

**Comment by guolinke** [comment]
just notice that you used `'feature_fraction': 0.8` (and `bagging_fraction`). there are some "randomnesses" in these sub-col/sub-row sampling. 
you can try to remove them.

**Comment by guolinke** [comment]
BTW, it is better to change the seed in the loop of the continued train, otherwise, the sampled cols/rows will be the same each time.

**Comment by CarlanLark** [comment]
Thanks a lot! After deleting `'feature_fraction'` and `'bagging_fraction'`, the gap becomes zero. By the word, if I want to use these two parameters, can I just build a random array for `'seed'` in params?

**Comment by guolinke** [comment]
@CarlanLark what is "random array for 'seed'"? 
The seed only can accept one integer value as input.
